test_runner/regress/test_replication_start_subxid_overflow.py

```diff
@@ -0,0 +1,254 @@
+
+
+def create_subxacts(endpoint, n_subxacts, delay):
+    with endpoint.connect() as con:
+        cur = con.cursor()
+        cur.execute("create table t(pk serial primary key, payload integer)")
+        cur.execute("begin")
+        cur.execute(
+            """create or replace function create_subxacts(n integer) returns void as $$
+                  declare
+                     i integer;
+                  begin
+                     for i in 1..n loop
+                        begin
+                           insert into t (payload) values (0);
+                        exception
+                           when others then
+                              raise exception 'caught something';
+                        end;
+                     end loop;
+                  end; $$ language plpgsql"""
+        )
+        cur.execute(f"select create_subxacts({n_subxacts})")
+        cur.execute("select txid_current()")
+        xid = cur.fetchall()[0][0]
+        log.info(f"Master transaction {xid}")
+        time.sleep(delay)
+        cur.execute("commit")
+
+
+# This test replica startup in case of large number of active subtransactions at primary node.
+# Number of known transaction xids is limited and if we do not limit number of transactions
+# restored from CLOG, then it will cause crash of replica.
+def test_replication_start_subxid_overflow(neon_simple_env: NeonEnv):
+    env = neon_simple_env
+
+    primary = env.endpoints.create_start(branch_name="main", endpoint_id="primary")
+    thread = threading.Thread(target=create_subxacts, args=(primary, 100000, 30))
+    thread.start()
+    secondary = env.endpoints.new_replica_start(origin=primary, endpoint_id="secondary")
+
+    time.sleep(20)
+    con = secondary.connect()
+    cur = con.cursor()
+
+    # Enforce setting hint bits for pg_class tuples.
+    # If master's transaction is not marked as in-progress in MVCC snapshot,
+    # then XMIN_INVALID hint bit will be set for table's 't' tuple makeing it invisible.
+    cur.execute("select * from pg_class")
+
+    # wait completion of transaction at primary
+    thread.join()
+    wait_replica_caughtup(primary, secondary)
+
+    cur.execute("select * from t where pk = 1")
+    assert cur.fetchone() == (1, 0)
+
+
+# Test starting a replica at a point in time where there is a transaction
+# running in primary with lots of subtransactions. Test MVCC in standby before
+# and after the commit of the transaction.
+#
+# XXX: This currently fails, the query in standby incorrectly sees some sub-xids
+# as committed:
+#
+# test_runner/regress/test_replication_start_subxid_overflow.py:105: AssertionError
+#
+# FAILED test_runner/regress/test_replication_start_subxid_overflow.py::test_replication_start_subxid_overflow2[debug-pg16] - assert (93229,) == (0,)
+@pytest.mark.skip("Failed to wait replica to start")
+def test_replication_start_subxid_overflow2(neon_simple_env: NeonEnv):
```

Comment:
Let's not skip this test.

Reply:
Ok, I will try to add interthread synchronisation
 to reenable this test.

Reply:
I saw only one possible way of enabling test_replication_start_subxid_overflow2 test: perform two different: once with overflown subxids and another with non0overflow xids. Fir this last one we can perform all this visibility tests/
@hlinnaka please look at the updated version of the test.

Concerning test_replication_start_subxid_overflow2 - I still do no see only two alternatives:
- either avoid "too many KnownAssignedXids" error (be reducing number of threads)
- either accept it